subsys/sd/sdio.c

```diff
@@ -56,23 +63,995 @@ int sdio_send_ocr(struct sd_card *card, uint32_t ocr)
+		card->flags |= SD_3000MV_FLAG;
+	}
+	/* Check mem present flag */
+	if (card->ocr & SDIO_OCR_MEM_PRESENT_FLAG) {
+		card->flags |= SD_MEM_PRESENT_FLAG;
+	}
+	/* Following steps are only required when using native SD mode */
+	if (IS_ENABLED(CONFIG_SDHC_SUPPORTS_NATIVE_MODE)) {
+		/*
+		 * If card and host support 1.8V, perform voltage switch sequence now.
+		 * note that we skip this switch if the UHS protocol is not enabled.
+		 */
+		if ((card->flags & SD_1800MV_FLAG) &&
+			(!card->host_props.is_spi) &&
+			(card->host_props.host_caps.vol_180_support) &&
+			IS_ENABLED(CONFIG_SD_UHS_PROTOCOL)) {
+			ret = sdmmc_switch_voltage(card);
+			if (ret) {
+				/* Disable host support for 1.8 V */
+				card->host_props.host_caps.vol_180_support = false;
+				/*
+				 * The host or SD card may have already switched to
+				 * 1.8V. Return SD_RESTART to indicate
+				 * negotiation should be restarted.
+				 */
+				card->status = CARD_ERROR;
+				return SD_RESTART;
+			}
+		}
+		if ((card->flags & SD_MEM_PRESENT_FLAG) &&
+			((card->flags & SD_SDHC_FLAG) == 0)) {
+			/* We must send CMD2 to get card cid */
+			ret = card_read_cid(card);
+			if (ret) {
+				return ret;
+			}
+		}
+		/* Send CMD3 to get card relative address */
+		ret = sdmmc_request_rca(card);
+		if (ret) {
+			return ret;
+		}
+		/* Move the card to transfer state (with CMD7) to run
+		 * remaining commands
+		 */
+		ret = sdmmc_select_card(card);
+		if (ret) {
+			return ret;
+		}
+	}
+	/* Read SDIO card common control register */
+	ret = sdio_read_cccr(card);
+	if (ret) {
+		return ret;
+	}
+	/* Initialize internal card function 0 structure */
+	card->func0.num = SDIO_FUNC_NUM_0;
+	card->func0.card = card;
+	ret = sdio_read_cis(&card->func0, cis_tuples,
+		ARRAY_SIZE(cis_tuples));
+	if (ret) {
+		return ret;
+	}
+
+	/* If card and host support 4 bit bus, enable it */
+	if (IS_ENABLED(CONFIG_SDHC_SUPPORTS_NATIVE_MODE) &&
+		((card->cccr_flags & SDIO_SUPPORT_HS) ||
+		(card->cccr_flags & SDIO_SUPPORT_4BIT_LS_BUS))) {
+		/* Raise bus width to 4 bits */
+		ret = sdio_set_bus_width(card, SDHC_BUS_WIDTH4BIT);
+		if (ret) {
+			return ret;
+		}
+		LOG_DBG("Raised card bus width to 4 bits");
+	}
+
+	/* Select and set bus speed */
+	sdio_select_bus_speed(card);
+	ret = sdio_set_bus_speed(card);
+	if (ret) {
+		return ret;
+	}
+	if (card->card_speed == SD_TIMING_SDR50 ||
+		card->card_speed == SD_TIMING_SDR104) {
+		/* SDR104, SDR50, and DDR50 mode need tuning */
+		ret = sdhc_execute_tuning(card->sdhc);
+		if (ret) {
+			LOG_ERR("SD tuning failed: %d", ret);
+		}
+	}
+	return ret;
+}
+
+
+/**
+ * @brief Enable SDIO function
+ *
+ * Enables SDIO card function
+ * @param func: function structure to initialize
+ * @param card: SD card to enable function on
+ * @param num: function number to initialize
+ * @retval 0 function was enabled successfully
+ * @retval -ETIMEDOUT: card I/O timed out
+ * @retval -EIO: I/O error
+ */
+int sdio_enable_func(struct sd_card *card, struct sdio_func *func,
```

Comment:
Can we split this one in to two?:
- Initialize SDIO function: `int sdio_init_func(struct sd_card *card, struct sdio_func *func, enum sdio_func_num num)`
- Enable SDIO function: 'int sdio_enable_func(struct sdio_func *func, enum sdio_func_num num)'

It will helpful for case as airoc-wifi driver, where enabling sdio functions is handling by low level wifi driver (WHD).
see my local changes of sdio.c in integration branch: https://github.com/npal-cy/zephyr-1/commit/50ad03a0cb9b1ec878ad6133014024cd99058ec8

Reply:
Can you clarify a bit here how splitting these functions would help with porting the AIROC driver? From what I can tell, the CIS issue you encountered will be resolved by the bugfix you suggested above (reading the CIS information from function 0).

Reply:
as i told before our WHD driver cares about enabling function F1, F2 in whd_bus_sdio_init according to device specific (e.g enabling F2 after FW downloaded to wifi device).  so if i try to enable F2 before WHD_Init, i will got timeout.

The spiting sdio_init_func/sdio_enable_func give me possibility to use CMD52/53 functions without enabling SDIO functions. Enabling SDIO functions will handling in WHD in proper sequences . WHD is separate asset, we do not want to modify it.

Ses my integration with using sdio_init_func: https://github.com/npal-cy/zephyr-1/commit/cba1a13b66b03b5126113aa67b316a7f51654a67

Reply:
Ok, thank you- I understand now. `whd_init` will leverage CMD52 and CMD53 to load the firmware, but SDIO functions cannot be enabled until you load the firmware. I'll split these functions into two sections, so that you can leverage the SDIO stack here fully.